Source/DafnyCore/AST/Attributes.cs

```diff
@@ -207,6 +238,240 @@ public static bool ContainsMatchingValue(Attributes attrs, string nm, ref object
+  public static Attributes ExpandAtAttribute(Program program, UserSuppliedAtAttribute atAttribute, IAttributeBearingDeclaration attributeHost) {
+    var toMatch = atAttribute.Arg;
+    var name = atAttribute.UserSuppliedName;
+    var bindings = atAttribute.UserSuppliedPreResolveBindings;
+
+    if (name == null) {
+      program.Reporter.Error(MessageSource.Resolver, atAttribute.RangeToken, "Attribute not recognized: " + atAttribute.ToString());
+      return null;
+    }
+
+    if (!TryGetBuiltinAtAttribute(name, out var builtinSyntax) || builtinSyntax == null) {
+      atAttribute.Builtin = false; // Will be resolved after
+      return null;
+    }
+
+    if (!builtinSyntax.CanBeApplied(attributeHost)) {
+      program.Reporter.Error(MessageSource.Resolver, atAttribute.RangeToken, UserSuppliedAtAttribute.AtName + atAttribute.UserSuppliedName + " attribute cannot be applied to " + attributeHost.WhatKind);
+    }
+
+    var formals = builtinSyntax.Args.Select(arg => arg.ToFormal()).ToArray();
+    ResolveLikeDatatypeConstructor(program, formals, name, atAttribute, bindings);
+
+    atAttribute.Builtin = true;
+    atAttribute.Arg.Type = Type.Int; // Dummy type to avoid crashes
+
+    switch (name) {
+      case "Compile": {
+          return A1("compile", bindings);
+        }
+      case "Fuel": {
+          if (Get(bindings, 0, out var lowFuel) && lowFuel != null) {
+            if (Get(bindings, 1, out var highFuel) && highFuel != null) {
+              if (Get(bindings, 2, out var functionName) && IsStringNotEmpty(functionName)) {
+                return A("fuel", functionName, lowFuel, highFuel);
+              }
+
+              return A("fuel", lowFuel, highFuel);
+            }
+
+            return A("fuel", lowFuel);
+          }
+
+          return A("fuel");
+        }
+      case "IsolateAssertions": {
+          return A("isolate_assertions");
+        }
+      case "Options": {
+          return A1("options", bindings);
+        }
+      default: {
+          throw new Exception("@-Attribute added to Attributes.BuiltinAtAttributes needs to be handled here");
+        }
+    }
+  }
+
+  // List of built-in @-attributes with their definitions.
+  // This list could be obtained from parsing and resolving a .Dfy file
+  // but for now it's good enough.
+  public static readonly List<BuiltInAtAttributeSyntax> BuiltinAtAttributes = new() {
+    BuiltIn("Compile")
+      .WithArg(TupleItem0Name, Type.Bool, DefaultBool(true))
+      .Filter(attributeHost =>
+          attributeHost is TopLevelDecl and not TypeParameter or MemberDecl or ModuleDefinition),
+    BuiltIn("Fuel")
+      .WithArg("low", Type.Int, DefaultInt(1))
+      .WithArg("high", Type.Int, DefaultInt(2))
+      .WithArg("functionName", Type.ResolvedString(), DefaultString(""))
+      .Filter(attributeHost => attributeHost is Function),
+    BuiltIn("IsolateAssertions")
+      .Filter(attributeHost => attributeHost is MemberDecl),
```

Comment:
Does this have an effect on Fields? Shouldn't it be `MethodOrFunction` instead?

Reply:
Yes it has an effect on constant fields with a RHS because the RHS might have some well formedness to be proved.
But I realize it should also support type with witnesses since there is a well-formedness check for them, and datatypes since parameters have default values. Same for newtypes.

Reply:
Consider changing to `attributeHost is ICanVerify`